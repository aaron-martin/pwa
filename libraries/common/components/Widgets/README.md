# Widgets
---

The `Widgets` component for displaying all widgets. It receives a `widgets` prop that it uses to build the individual widgets and maps them to the correct components.

> **Dependencies:**  
> `<Grid />`, `<Widget />`

## Getting Started

```jsx
import Widgets from '@shopgate/pwa-common/components/Widgets';

<Widgets background="orange" widgets={...} />
```

## Props

### background

_Type_: `string`  

The `background` prop will determine the background of the widget area.

### widgets

_Type_: `Object`  

The `widgets` prop contains all of the necessary information to build the widgets.
---
