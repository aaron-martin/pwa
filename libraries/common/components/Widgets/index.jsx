/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import WidgetGrid from './components/WidgetGrid';

const WIDGET_GRID_TYPE = '@shopgate/commerce-widgets/widget-grid';
const GRID_COLUMNS = 12; // One grid row has 12 columns.

/**
 * Creates a grid wrapper for widget(s).
 * @param {string} key The unique key.
 * @param {Array} config Array of widgets to be placed in the grid.
 * @param {Array} components The component definitions for the widgets.
 * @returns {JSX} The wrapper.
 */
const createGridWrapper = (key, config, components) => (
  React.createElement(
    WidgetGrid,
    {
      key,
      config,
      components,
    }
  )
);

/**
 * Create array of elements from widget configuration.
 * @param {Array} widgets Array of widget configurations.
 * @param {Array} components The component definitions for the widgets.
 * @returns {Array} Array of JSX elements.
 */
const createArrayOfElements = (widgets, components) => (
  (widgets || []).map((widget, index) => {
    if (!components[widget.type] && widget.type !== WIDGET_GRID_TYPE) {
      return null;
    }

    const key = `w${index}`;

    if (widget.type === WIDGET_GRID_TYPE) {
      // If it's a grid just create it and pass the child widgets.
      return createGridWrapper(key, widget.settings.widgets, components);
    } else if (widget.height) {
      // If it has a definite height wrap the widget in a grid.
      return createGridWrapper(
        key,
        [{
          ...widget,
          col: 0,
          row: 0,
          width: GRID_COLUMNS,
        }],
        components
      );
    }

    // In all other cases just create and return the widget component.
    return React.createElement(
      components[widget.type],
      {
        ...widget,
        key,
      }
    );
  })
);

/**
 * The Widgets component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const Widgets = ({ widgets, components }) => {
  if (!widgets) {
    return null;
  }

  return (
    <div>{createArrayOfElements(widgets, components)}</div>
  );
};

Widgets.propTypes = {
  components: PropTypes.shape(),
  widgets: PropTypes.arrayOf(PropTypes.shape()),
};

Widgets.defaultProps = {
  components: null,
  widgets: null,
};

export default Widgets;
