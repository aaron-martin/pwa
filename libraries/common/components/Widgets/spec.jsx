/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Widgets from './index';

/**
 * A mock Image component.
 * @returns {JSX}
 */
const Image = () => <div />;

const components = {
  '@shopgate/commerce-widgets/image': Image,
};

describe('<Widgets />', () => {
  it('should render a grid if height is defined', () => {
    const widgets = [{
      col: 0,
      row: 0,
      width: 12,
      height: 3,
      settings: {
        id: 83535,
        image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
      },
      type: '@shopgate/commerce-widgets/image',
    }];

    const wrapper = mount(
      <Widgets
        components={components}
        widgets={widgets}
      />
    );

    expect(wrapper.find('WidgetGrid').exists()).toBe(true);
  });

  it('should not wrap a widget which is not a grid and has no height', () => {
    const widgets = [{
      col: 0,
      row: 0,
      width: 12,
      settings: {
        id: 83535,
        image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
      },
      type: '@shopgate/commerce-widgets/image',
    }];

    const wrapper = mount(
      <Widgets
        components={components}
        widgets={widgets}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('WidgetGrid').exists()).toBe(false);
  });

  it('should render a grid if the widget is of type grid', () => {
    const widgets = [{
      type: '@shopgate/commerce-widgets/widget-grid',
      settings: {
        widgets: [
          {
            col: 0,
            row: 0,
            width: 12,
            height: 5,
            settings: {
              id: '84961',
              alt: '',
              image: 'https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg',
              link: '/category/',
            },
            type: '@shopgate/commerce-widgets/image',
          },
        ],
      },
    }];

    const wrapper = mount(
      <Widgets
        components={components}
        widgets={widgets}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('WidgetGrid').exists()).toBe(true);
  });
});
