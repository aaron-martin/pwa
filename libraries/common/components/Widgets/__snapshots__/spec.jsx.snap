// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Widgets /> should not wrap a widget which is not a grid and has no height 1`] = `
<Widgets
  components={
    Object {
      "@shopgate/commerce-widgets/image": [Function],
    }
  }
  widgets={
    Array [
      Object {
        "col": 0,
        "row": 0,
        "settings": Object {
          "id": 83535,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        },
        "type": "@shopgate/commerce-widgets/image",
        "width": 12,
      },
    ]
  }
>
  <div>
    <Image
      col={0}
      key="w0"
      row={0}
      settings={
        Object {
          "id": 83535,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        }
      }
      type="@shopgate/commerce-widgets/image"
      width={12}
    >
      <div />
    </Image>
  </div>
</Widgets>
`;

exports[`<Widgets /> should render a grid if the widget is of type grid 1`] = `
<Widgets
  components={
    Object {
      "@shopgate/commerce-widgets/image": [Function],
    }
  }
  widgets={
    Array [
      Object {
        "settings": Object {
          "widgets": Array [
            Object {
              "col": 0,
              "height": 5,
              "row": 0,
              "settings": Object {
                "alt": "",
                "id": "84961",
                "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
                "link": "/category/",
              },
              "type": "@shopgate/commerce-widgets/image",
              "width": 12,
            },
          ],
        },
        "type": "@shopgate/commerce-widgets/widget-grid",
      },
    ]
  }
>
  <div>
    <WidgetGrid
      components={
        Object {
          "@shopgate/commerce-widgets/image": [Function],
        }
      }
      config={
        Array [
          Object {
            "col": 0,
            "height": 5,
            "row": 0,
            "settings": Object {
              "alt": "",
              "id": "84961",
              "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
              "link": "/category/",
            },
            "type": "@shopgate/commerce-widgets/image",
            "width": 12,
          },
        ]
      }
      key="w0"
    >
      <div
        className="css-13azwyo"
        style={
          Object {
            "height": "425px",
          }
        }
      >
        <Widget
          cellSize={85}
          component={[Function]}
          config={
            Object {
              "col": 0,
              "height": 5,
              "row": 0,
              "settings": Object {
                "alt": "",
                "id": "84961",
                "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
                "link": "/category/",
              },
              "type": "@shopgate/commerce-widgets/image",
              "width": 12,
            }
          }
          key="w0"
        >
          <GridItem
            className="css-147ej5f css-63oe3q  "
            component="div"
            grow={0}
            shrink={1}
          >
            <div
              className="css-147ej5f css-63oe3q  "
            >
              <div
                className="css-4hpmhj"
              >
                <Image
                  ratio={
                    Array [
                      12,
                      5,
                    ]
                  }
                  settings={
                    Object {
                      "alt": "",
                      "id": "84961",
                      "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
                      "link": "/category/",
                    }
                  }
                >
                  <div />
                </Image>
              </div>
            </div>
          </GridItem>
        </Widget>
      </div>
    </WidgetGrid>
  </div>
</Widgets>
`;
